### PR TITLE
drivers: sensor: expanding driver for STM32 quadrature encoder

### PR DESCRIPTION
**Pull request by peppedxAlto**
Added a mode node to the qdec bindings.
And enable qdec for TIM2 and TIM3 of STM32F3.

### CODE REVIEW

File: samples/sensor/qdec/boards/nucleo_f303re.overlay [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    aliases {
+        qdec0 = &qdec;
+    };
+};
+
+
+/ { chosen {
+        zephyr,canbus = &can1;
+    };
+};
```
Comment (erwango) [inline samples/sensor/qdec/boards/nucleo_f303re.overlay:17]: Doesn't seem related

File: include/zephyr/dt-bindings/sensor/qdec_stm32.h [hunk 2]
```diff
@@ -23,4 +23,8 @@
 #define FDIV32_N6	14
 #define FDIV32_N8	15
 
+#define MODE_X2_TI1  0
+#define MODE_X2_TI2  1
+#define MODE_X4_TI12 2
```
Comment (erwango) [inline include/zephyr/dt-bindings/sensor/qdec_stm32.h:28]: Non blocking: adjust indentation to fit previous lines

File: dts/bindings/sensor/st,stm32-qdec.yaml [hunk 3]
```diff
@@ -70,3 +70,13 @@ properties:
...
       were done based on the current counter's value.
+
+  st,encoder-mode:
+    type: int
+    description: Encoder counting mode
+    default: 0
+    enum:
+      - 0 # MODE_X2_TI1 Counter counts up/down on TI2FP2 edge depending on TI1FP1 level
+      - 1 # MODE_X2_TI2  Counter counts up/down on TI1FP1 edge depending on TI2FP2 level
+      - 2 # MODE_X4_TI12 Counter counts up/down on both TI1FP1 and TI2FP2 edges
```
Comment (erwango) [inline dts/bindings/sensor/st,stm32-qdec.yaml:81]: Please move details into description, otherwise  they won't be displayed in doc.
See https://docs.zephyrproject.org/latest/build/dts/api/bindings/sensor/st%2Cstm32-qdec.html#dtbinding-st-stm32-qdec

File: drivers/sensor/qdec_stm32/qdec_stm32.c [hunk 4]
```diff
@@ -25,13 +25,18 @@
 
 LOG_MODULE_REGISTER(qdec_stm32, CONFIG_SENSOR_LOG_LEVEL);
 
+#define MODE_X2_TI1  0
+#define MODE_X2_TI2  1
+#define MODE_X4_TI12 2
```
Comment (erwango) [inline drivers/sensor/qdec_stm32/qdec_stm32.c:30]: Why redefining those here ?

File: drivers/sensor/qdec_stm32/qdec_stm32.c [hunk 5]
```diff
@@ -150,6 +167,7 @@ static const struct sensor_driver_api qdec_stm32_driver_api = {
 		.is_input_polarity_inverted = DT_INST_PROP(n, st_input_polarity_inverted),	\
 		.input_filtering_level = DT_INST_PROP(n, st_input_filter_level),		\
 		.counts_per_revolution = DT_INST_PROP(n, st_counts_per_revolution),		\
+		.encoder_mode = DT_INST_PROP(n, st_encoder_mode),  \
```
Comment (erwango) [inline drivers/sensor/qdec_stm32/qdec_stm32.c:170]: Nit, onn blocking: fix end of line indentation

File: samples/sensor/qdec/boards/nucleo_f303re.overlay [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2022 Giuseppe Iellamo <giuseppe.iellamo@altorobotics.ai>
```
Comment (erwango) [inline samples/sensor/qdec/boards/nucleo_f303re.overlay:2]:
```suggestion
 * Copyright 2023 Giuseppe Iellamo <giuseppe.iellamo@altorobotics.ai>
```

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/sensor/qdec_stm32/qdec_stm32.c (hunk 4 above)
Comment (peppedxAlto) [reply drivers/sensor/qdec_stm32/qdec_stm32.c:30]: I don´t like but I wasn´t sure if it was good practicse to include stuff from the bindings.

File: samples/sensor/qdec/boards/nucleo_f303re.overlay (hunk 1 above)
Comment (peppedxAlto) [reply samples/sensor/qdec/boards/nucleo_f303re.overlay:17]: Ok

File: include/zephyr/dt-bindings/sensor/qdec_stm32.h (hunk 2 above)
Comment (peppedxAlto) [reply include/zephyr/dt-bindings/sensor/qdec_stm32.h:28]: Not sure if I have understood; do you mean aligning the numbers?

File: samples/sensor/qdec/boards/nucleo_f303re.overlay (hunk 6 above)
Comment (peppedxAlto) [reply samples/sensor/qdec/boards/nucleo_f303re.overlay]: I pushed the fix

File: include/zephyr/dt-bindings/sensor/qdec_stm32.h (hunk 2 above)
Comment (peppedxAlto) [reply include/zephyr/dt-bindings/sensor/qdec_stm32.h:28]: Oh I understood I was using tab expansion and seeing things differently.

Sorry.

File: dts/arm/st/f3/stm32f3.dtsi
Comment (erwango) [inline dts/arm/st/f3/stm32f3.dtsi:1]: This change isn't directly related to driver modification and should be put in a dedicated commit.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for the update, and this contribution in general.
One last comment, otherwise LGTM

File: dts/arm/st/f3/stm32f3.dtsi
Comment (peppedxAlto) [reply dts/arm/st/f3/stm32f3.dtsi:1]: Will it suffice a dedicated commit in this PR or should I remove it entirely ? 
(and with it also the nucleo_f303re.overlay file)

Comment (erwango) [reply dts/arm/st/f3/stm32f3.dtsi:1]: In this PR this is ok

File: dts/bindings/sensor/st,stm32-qdec.yaml [hunk 7]
```diff
@@ -70,3 +70,17 @@ properties:
...
+
+  st,encoder-mode:
+    type: int
+    description: |
+      Encoder counting mode
+      MODE_X2_TI1 Counter counts up/down on TI2FP2 edge depending on TI1FP1 level
+      MODE_X2_TI2  Counter counts up/down on TI1FP1 edge depending on TI2FP2 level
+      MODE_X4_TI12 Counter counts up/down on both TI1FP1 and TI2FP2 edges
+      depending on the level of the other input
+    default: 0
```
Comment (MaureenHelm) [inline dts/bindings/sensor/st,stm32-qdec.yaml:82]: The default needs a justification in the description. Is it a register reset value?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/st,stm32-qdec.yaml (hunk 7 above)
Comment (peppedxAlto) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: It is the actual default value of the driver and of the STM32 LL.

Comment (peppedxAlto) [reply dts/bindings/sensor/st,stm32-qdec.yaml:83]: I've added to the description.

File: dts/bindings/sensor/st,stm32-qdec.yaml [hunk 8]
```diff
@@ -70,3 +70,18 @@ properties:
...
+
+  st,encoder-mode:
+    type: int
+    description: |
+      Encoder counting mode
+      MODE_X2_TI1 Counter counts up/down on TI2FP2 edge depending on TI1FP1 level
+      MODE_X2_TI2  Counter counts up/down on TI1FP1 edge depending on TI2FP2 level
+      MODE_X4_TI12 Counter counts up/down on both TI1FP1 and TI2FP2 edges
+      depending on the level of the other input
+      Defaults on MODE_X2_TI1 since it is the low level driver default
```
Comment (erwango) [inline dts/bindings/sensor/st,stm32-qdec.yaml:82]: Doesn't look like a right justification (and for sure not to the reader).
Isn't it the HW default reset value ?

Comment (peppedxAlto) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: No after reset encoder mode is not enabled. You have to program if you want to use encoders.
The default I chose is 
1- The actual behavior of the driver already in mainline
2- The default of the Low Level Driver.

Comment (erwango) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: Btw, here is the reason why we're bothering you with this: https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

In that case, setting a default value wouldn't break anything, but I'd say that this is now an application choice, with no reason to default to one or the other (except the legacy behavior).

The only good reason would be that with this change, not providing default would break existing users. But I think we should rather try to detect if missing and raise an error. Setting it as `required: true` would allow to do that, and I think that should be sufficient. Since it would be a breaking change, this would require to be documented in release notes.

@MaureenHelm What would be you opinion?

Comment (peppedxAlto) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: I see. This is my first attempt at a contribution and I am trying to do it because I thought it was a good idea to give back this super tiny bit; that said I tried to follow the guidelines but since I am doing it in my spare time I haven´t studied all the possible documents.

Now I am leaving for a couple of weeks, so I won't be able to follow up, so if you want to shelve this no problem, maybe I will try again in the future.

Comment (MaureenHelm) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: The default value in the binding should correspond to the hardware default. It's unfortunate that the LL driver default doesn't match, so in this case I think it makes sense to make the property required to avoid quietly changing driver behavior.

Comment (peppedxAlto) [reply dts/bindings/sensor/st,stm32-qdec.yaml]: Hi @erwango @MaureenHelm I pushed the modification to change the field from defaulted to required.

File: dts/bindings/sensor/st,stm32-qdec.yaml [hunk 9]
```diff
@@ -70,3 +70,17 @@ properties:
...
+
+  st,encoder-mode:
+    type: int
+    description: |
+      Encoder counting mode
+      MODE_X2_TI1 Counter counts up/down on TI2FP2 edge depending on TI1FP1 level
+      MODE_X2_TI2  Counter counts up/down on TI1FP1 edge depending on TI2FP2 level
+      MODE_X4_TI12 Counter counts up/down on both TI1FP1 and TI2FP2 edges
+      depending on the level of the other input
+    required: true
```
Comment (erwango) [inline dts/bindings/sensor/st,stm32-qdec.yaml:82]: Since this is a breaking change for existing users, please update `doc/releases/migration-guide-3.5.rst` (as a required change)

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/st,stm32-qdec.yaml (hunk 9 above)
Comment (peppedxAlto) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: > * Quadrature decoder STM32 driver now has a new property :dtcompatible:`st,encoder-mode` which is required.
  The old behavior is the same as setting the property as :c:enumerator:`MODE_X2_TI1`.
  
Will this kind of text suffice?

Comment (erwango) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: Perfect!

Comment (MaureenHelm) [reply dts/bindings/sensor/st,stm32-qdec.yaml:82]: This change should be squashed into the commit that introduces the `st,encoder-mode` property

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
